test/models/user_test.rb

```diff
@@ -4,4 +4,35 @@ class UserTest < ActiveSupport::TestCase
   # test "the truth" do
   #   assert true
   # end
+
+  EMAIL_REGEX = /\A[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,4}\Z/i
+
+  def setup
+    @user = User.create(username: "User1", email: "user1sfu.ca", password: "secrete")
+  end
+
+  test "should not have the same username" do
+    duplicate_user = @user.dup
+    assert_not duplicate_user.valid?, "Create the user with existing username"
+  end
+
+  test "should not have the same email" do
+    duplicate_user = @user.dup
+    assert_not duplicate_user.valid?, "Create the user with existing email"
+  end
+
+  test "should not have blank user name" do
+    user = User.new(password: "secrete")
+    assert_not user.save, "Save the user without an username"
+  end
+
+  test "should not have blank email" do
+    user = User.new(password: "secrete")
+    assert_not user.save, "Save the user without an emal"
+  end
+
+  test "should not have invalid email format" do
+    assert_match(EMAIL_REGEX, @user.email, "Invalid email format")
```

Comment:
@user.email is "user1sfu.ca". Wouldn't this assertion always fail? 

Besides, this test should be in the same format as the blank email, blank username tests. validation of email format should happen on the model https://apidock.com/rails/ActiveModel/Validations/ClassMethods/validates_format_of

perhaps we can create two users in this test, 1 with valid email format and 1 with an invalid one, and assert that the 1 with valid email format is saved successfully where as the other one is not.

Reply:
Yeap. Validation for email format should be provided in the model. I am thinking how to write test for it? We can save and assert it but then when we run test, it will not allow us to assert_match.

Reply:
perhaps something like
assert (EMAIL_REGEX.match(user.email) && user.save)
assert_not (user.save && !EMAIL_REGEX.match(user.email))

not too familiar with rails assertion syntax.